sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeJoinToEmptyRelation.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.planning.ExtractSingleColumnNullAwareAntiJoin
+import org.apache.spark.sql.catalyst.plans.{Inner, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.joins.{EmptyHashedRelation, EmptyHashedRelationWithAllNullKeys, HashedRelation}
+
+/**
+ * This optimization rule detects and converts a `Join` to an empty `LocalRelation`:
+ * 1. `Join` is single column NULL-aware anti join (NAAJ), and broadcasted `HashedRelation`
+ *    is `EmptyHashedRelationWithAllNullKeys`.
+ *
+ * 2. `Join` is inner or left semi join, and broadcasted `HashedRelation` is `EmptyHashedRelation`.
```

Comment:
how about other join types?

Reply:
like left outer?

Reply:
@cloud-fan - for left outer join, we need to output the stream side (left side) right? Not sure how AQE helps here.

Reply:
left outer with left empty should return nothing?

Reply:
@cloud-fan - with current SHJ/BHJ implementation, left outer join can only build right side, but not left side. I want to target this PR to only handle empty hashed relation (empty build side). But I agree, we can have a short-cut to check stream side is empty or not, before building hashed relation, to save the cost of building hashed relation. But I think AQE cannot handle it, and we can tackle it separately if it's worthy. WDYT?

Reply:
you are right, let's leave it for later.